### PR TITLE
Add laser retro support in Ogre2

### PR DESCRIPTION
**Pull request by doisyg**
As discussed here https://github.com/ignitionrobotics/ign-rendering/issues/181 and suggested by @iche033 . Taking huge inspiration and copypasting from Ogre2ThermalCamera (not sure I get the full shader logic). The implementation here is similar as the temperature tag from the Ogre2ThermalCamera.
laser retro values are valid on the range 0.0 to 2000.0 as it was, I believe, on Gazebo Classic. It can be an opportunity to discuss and change this behavior if needed.
There seem to be a loss of precision of around 5 retro value (see the test), don't know if it is due to some noise added in the pipeline or some shader witchery, but it works well for my use case (simulating lidar reflective tape).
With this PR, a pending PR in sdformat (https://github.com/osrf/sdformat/pull/454) and one in ign-gazebo that I will soon post (but depends on sdformat), laser retro value will be settable from a sdf/urdf file.

### CODE REVIEW

**Review by iche033 — COMMENTED** [review]
nice, thanks for the contribution!

I'm curious where you find the 2000 limit in gazebo? I was looking through gazebo code but couldn't see where that limit is set.

We ran into precision issue with thermal camera readings as well. Need to investigate further to find out how to resolve this

File: ogre2/src/Ogre2GpuRays.cc [hunk 1]
```diff
@@ -99,11 +158,132 @@ class ignition::rendering::Ogre2GpuRaysPrivate
...
+              retro_value = -1.0;
+            }
+          }
+        }
+
+        // only accept positive laser retro value
+        if (retro_value >= 0)
+        {
+          // set visibility flag so the camera can see it
+          item->addVisibilityFlags(0x10000000);
```
[the remark this answers is not in the document]
Comment (doisyg) [reply ogre2/src/Ogre2GpuRays.cc:249]: It seems to work, and if I removed it, the I get 0 retro values

File: ogre2/src/Ogre2GpuRays.cc [hunk 2]
```diff
@@ -589,6 +820,24 @@ void Ogre2GpuRays::Setup1stPass()
     this->dataPtr->ogreCompositorWorkspace1st[i] =
         ogreCompMgr->addWorkspace(this->scene->OgreSceneManager(),
         rt, this->dataPtr->cubeCam[i], wsDefName, false);
+
+    // add laser retro material switcher to render target listener
+    // so we can switch to use laser retro material when the camera is being udpated
+    Ogre::CompositorNode *node =
+        this->dataPtr->ogreCompositorWorkspace1st[i]->getNodeSequence()[0];
+    auto channels = node->getLocalTextures();
```
[the remark this answers is not in the document]
Comment (doisyg) [reply ogre2/src/Ogre2GpuRays.cc:828]: Renamed the variable

File: ogre2/src/Ogre2GpuRays.cc [hunk 3]
```diff
@@ -493,7 +686,26 @@ void Ogre2GpuRays::Setup1stPass()
...
+    colorTexDef->fsaa = 0;
+    colorTexDef->uav = false;
+    colorTexDef->automipmaps = false;
+    colorTexDef->hwGammaWrite = Ogre::TextureDefinitionBase::BoolFalse;
+    colorTexDef->depthBufferId = Ogre::DepthBuffer::POOL_DEFAULT;
+    colorTexDef->depthBufferFormat = Ogre::PF_D32_FLOAT;
+    colorTexDef->preferDepthTexture = true;
+    colorTexDef->fsaaExplicitResolve = false;
+
+    nodeDef->setNumTargetPass(3);
```
[the remark this answers is not in the document]
Comment (doisyg) [reply ogre2/src/Ogre2GpuRays.cc:708]: Note sure I fully understand it but your path works

File: ogre2/src/Ogre2GpuRays.cc [hunk 4]
```diff
@@ -99,11 +158,132 @@ class ignition::rendering::Ogre2GpuRaysPrivate
...
+        {
+          // set visibility flag so the camera can see it
+          item->addVisibilityFlags(0x10000000);
+          for (unsigned int i = 0; i < item->getNumSubItems(); ++i)
+          {
+            Ogre::SubItem *subItem = item->getSubItem(i);
+            if (!subItem->hasCustomParameter(this->customParamIdx))
+            {
+              // limit laser retro value to 2000 (as in gazebo)
+              if (retro_value > 2000.0){
```
[the remark this answers is not in the document]
Comment (doisyg) [reply ogre2/src/Ogre2GpuRays.cc]: See discussion below

**Comment by doisyg** [comment]
> I'm curious where you find the 2000 limit in gazebo? I was looking through gazebo code but couldn't see where that limit is set.

You're right, there is no limit in gazebo classic, I somehow remembered that the max was 2000.0. I know for sure that I never encountered a laser_retro value (or ROS laserscan intensityhttp://docs.ros.org/en/api/sensor_msgs/html/msg/LaserScan.html) above 2000. But since the intensity is a lidar vendor specific unit maybe we should left it unbounded. What do you think ?

> We ran into precision issue with thermal camera readings as well. Need to investigate further to find out how to resolve this

For my use case, since lidar intensity measurement are always noisy, this is fine.

File: ogre2/src/Ogre2GpuRays.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (iche033) [reply ogre2/src/Ogre2GpuRays.cc:249]: I was thinking of using a different value, eg. 0x01000000

You'll also need to update [this line](https://github.com/ignitionrobotics/ign-rendering/blob/60b4f54df1f0464a5fc943fee12393651d01a717/ogre2/src/Ogre2GpuRays.cc#L725) to match.

**Review by iche033 — COMMENTED** [review]
> But since the intensity is a lidar vendor specific unit maybe we should left it unbounded. What do you think ?

yeah I agree. I think we should make it unbounded.

**Comment by doisyg** [comment]
> yeah I agree. I think we should make it unbounded.

Thinking a bit more about making it unbounded, I think we have a non trivial issue about the range we can accept due to the need of normalizing the value to [0,1] on a float to pass it to the pipeline:
https://github.com/ignitionrobotics/ign-rendering/blob/31af089b83d9dc23d1c46ed94cf258dd0e294fdc/ogre2/src/Ogre2GpuRays.cc#L259-L261

Float precision is not constant within its range: https://blog.demofox.org/2017/11/21/floating-point-precision/
On [0, 1], I believe can assume a precision of at least `0.00000011920929`

Also, what do we want as a minimal laser_retro  value step ? 0.1 ? 0.01 ?

**Comment by iche033** [comment]
thanks for looking into the precision issue. Sounds like we still need an upper bound.

> Also, what do we want as a minimal laser_retro value step ? 0.1 ? 0.01 ?

I tried to play around with different upper bound vs small retro values and noticed that we can't really have small step sizes due to the the precision / noise issue which you also noticed in your test. So I would not worry about trying to support a small step size for now until we fixed this problem.

**Comment by doisyg** [comment]
> I tried to play around with different upper bound vs small retro values and noticed that we can't really have small step sizes due to the the precision / noise issue which you also noticed in your test. So I would not worry about trying to support a small step size for now until we fixed this problem.

Ok, I guess leaving it bounded to 2000.0 for now should be ok (and easily changeable to another value if needed).
Any other blocking points ?

File: ogre2/src/Ogre2GpuRays.cc [hunk 5]
```diff
@@ -439,6 +620,7 @@ void Ogre2GpuRays::Setup1stPass()
   // {
   //   in 0 rt_input
   //   texture depthTexture target_width target_height PF_D32_FLOAT
+  //   texture colorTexture target_width target_height PF_R8G8B8
   //   target depthTexture
```
Comment (iche033) [inline ogre2/src/Ogre2GpuRays.cc:624]: since we removed the depthTexture target pass in this PR, you can remove this part of the comment.

**Review by iche033 — COMMENTED** [review]
looks good to me. Can you fix DCO?

File: ogre2/src/Ogre2GpuRays.cc (hunk 5 above)
Comment (doisyg) [reply ogre2/src/Ogre2GpuRays.cc:624]: Did you mean this line ? :
https://github.com/ignitionrobotics/ign-rendering/blob/6f27238678eb71567d698c68430a5ec8f56db9f7/ogre2/src/Ogre2GpuRays.cc#L622

And I guess this line too ?:
https://github.com/ignitionrobotics/ign-rendering/blob/6f27238678eb71567d698c68430a5ec8f56db9f7/ogre2/src/Ogre2GpuRays.cc#L653

and modifying https://github.com/ignitionrobotics/ign-rendering/blob/6f27238678eb71567d698c68430a5ec8f56db9f7/ogre2/src/Ogre2GpuRays.cc#L654 to `input 0 colorTexture`

(I am not familiar with the shader syntax)

Comment (iche033) [reply ogre2/src/Ogre2GpuRays.cc:624]: we're still creating the depth texture and feeding that to the `rt_input` target pass. But we are no longer rendering the `depthTexture` target pass. So we can remove these lines: https://github.com/ignitionrobotics/ign-rendering/blob/6f27238678eb71567d698c68430a5ec8f56db9f7/ogre2/src/Ogre2GpuRays.cc#L624-L633

**Comment by doisyg** [comment]
Fixed DCO mess by re-comiting from scratch

**Comment by doisyg** [comment]
Anything else blocking the merge ?

**Review by iche033 — APPROVED** [review]
CI looks good. Approving.

**Pull request merged**